RSSReader/View/FeedViewController.m

```diff
@@ -73,38 +98,66 @@ - (UITableView *)tableView {
         _tableView.translatesAutoresizingMaskIntoConstraints = NO;
         _tableView.dataSource = self;
         _tableView.delegate = self;
-        [_tableView registerClass:FeedCell.class forCellReuseIdentifier:@"cellId"];
+        [_tableView registerClass:FeedCell.class forCellReuseIdentifier:kCellID];
     }
     return _tableView;
 }
 
+- (NSMutableDictionary *)cellHeightsDictionary {
+    if (!_cellHeightsDictionary) {
+        _cellHeightsDictionary = [NSMutableDictionary new];
+    }
+    return _cellHeightsDictionary;
+}
+
+#pragma mark TableViewDataSourse implementation
 
 - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
-    FeedCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellId"];
-    if (self.feedItemArray.count != 0) {
-        [cell setItem: self.feedItemArray[indexPath.row]];
-    }
-    [cell setupCell];
+    FeedCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellID];
+    [cell configureWithItem: self.feedItemArray[indexPath.row] index:(int)indexPath.row];
+    cell.delegate = self;
     return cell;
 }
 
 - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
     return self.feedItemArray.count;
 }
 
-- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
-    return UITableViewAutomaticDimension;
-}
+#pragma mark TableViewDelegate implementation
 
 - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
     [tableView deselectRowAtIndexPath:indexPath animated:YES];
-    [[UIApplication sharedApplication] openURL:[[[NSURL alloc] initWithString:self.feedItemArray[indexPath.row].link] autorelease]];
+    NSURL *url = [NSURL URLWithString:self.feedItemArray[indexPath.row].link];
+    WebViewController *webViewController = [[WebViewController alloc] initWithURL:url];
+    
+    [self.navigationController pushViewController:webViewController animated:YES];
+    [webViewController release];
+}
+
+- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
+    [self.cellHeightsDictionary setObject:@(cell.frame.size.height) forKey:indexPath];
+}
+
+- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath {
+    NSNumber *height = [self.cellHeightsDictionary objectForKey:indexPath];
+    return [height boolValue] ? height.doubleValue : UITableViewAutomaticDimension;
```

Comment:
That looks like a workaround. Why do you use UITableViewAutomaticDimension and keep setting cell height manually???

Reply:
I ran into bug with cell animation. So this is the only way that helped me fix it.
I tried to fix it in different ways, but this is the only solution

Reply:
Did you check it diffable data source?

Reply:
yes, but implementation of DiffableDataSourse in context of objective-C raises too many questions. I decided to use this way, because it very short and easy. In context of Swift I will use diffable data sourse